[Writing Tools] Refactor Writing Tools type conversion methods to be in their own file

rr-codes opened the pull request:
#### 0af20f269c8d2afd0298c7200ae15b25124a3065
<pre>
[Writing Tools] Refactor Writing Tools type conversion methods to be in their own file
<a href="https://bugs.webkit.org/show_bug.cgi?id=275599">https://bugs.webkit.org/show_bug.cgi?id=275599</a>
<a href="https://rdar.apple.com/130054970">rdar://130054970</a>

Reviewed by Aditya Keerthi.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView wantsWritingToolsInlineEditing]):
(-[WKWebView willBeginWritingToolsSession:requestContexts:]):
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView proofreadingSession:didReceiveSuggestions:processedRange:inContext:finished:]):
(-[WKWebView proofreadingSession:didUpdateState:forSuggestionWithUUID:inContext:]):
(-[WKWebView didEndWritingToolsSession:accepted:]):
(-[WKWebView compositionSession:didReceiveText:replacementRange:inContext:finished:]):
(-[WKWebView writingToolsSession:didReceiveAction:]):
(-[WKWebView _textReplacementSession:updateState:forReplacementWithUUID:]):
(convert): Deleted.
(-[WKWebView _wantsCompleteUnifiedTextReplacementBehavior]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration setWritingToolsBehavior:]):
(-[WKWebViewConfiguration writingToolsBehavior]):
(convertToPlatform): Deleted.
(convertToWeb): Deleted.
(-[WKWebViewConfiguration _setUnifiedTextReplacementBehavior:]): Deleted.
(-[WKWebViewConfiguration _unifiedTextReplacementBehavior]): Deleted.
(convert): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _web_wantsCompleteUnifiedTextReplacementBehavior]):
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h: Added.
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.mm: Added.
(WebKit::convertToPlatformWritingToolsBehavior):
(WebKit::convertToPlatformTextSuggestionState):
(WebKit::convertToPlatformContext):
(WebKit::convertToWebWritingToolsBehavior):
(WebKit::convertToWebTextSuggestionState):
(WebKit::convertToWebAction):
(WebKit::convertToWebSessionType):
(WebKit::convertToWebCompositionSessionType):
(WebKit::convertToWebContext):
(WebKit::convertToWebSession):
(WebKit::convertToWebTextSuggestion):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::selectionDidChange):
(WebKit::WebViewImpl::canHandleSwapCharacters const):
(WebKit::WebViewImpl::handleContextMenuSwapCharacters):
(WebKit::showSwapCharactersViewRelativeToRectOfView): Deleted.
(WebKit::scheduleShowSwapCharactersViewForSelectionRectOfView): Deleted.
(WebKit::webViewCanHandleSwapCharacters): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/280134@main">https://commits.webkit.org/280134@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d23bb3fbdfc31f4632c4b640cf2f912f2b74d62

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55844 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58840 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/6276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57970 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42789 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6472 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/44968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/6276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57873 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/33064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/48141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/26101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29848 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/5469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4419 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/5740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/60429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/5867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/60429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/48210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/60429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8238 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/30997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32081 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/33163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/31829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2207,78 +2092,79 @@ - (UIWritingToolsAllowedInputOptions)writingToolsAllowedInputOptions {
+#if PLATFORM(IOS)
+    return [self _isEditable] || [_configuration writingToolsBehavior] == UIWritingToolsBehaviorComplete;
+#elif PLATFORM(MAC)
+    return [self _isEditable] || [_configuration writingToolsBehavior] == NSWritingToolsBehaviorComplete;
```
Can we also use your utilities file to unify `UIWritingToolsBehaviorComplete` / `NSWritingToolsBehaviorComplete`?

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h:
```diff
@@ -0,0 +1,84 @@
+
+#if ENABLE(WRITING_TOOLS)
+
+#import <pal/spi/cocoa/WritingToolsSPI.h>
```
I think you should be able to forward declare the classes you need.

pxlcoder replied:
Oh but maybe not the enums?

rr-codes replied:
ya, I wasn't sure what to do about the enums

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -2802,7 +2794,8 @@ static String commandNameForSelector(SEL selector)
 
-        scheduleShowSwapCharactersViewForSelectionRectOfView(selectionRect, m_view.getAutoreleased());
+        // The affordance will only show up if the selected range consists of >= 50 characters.
+        [WTWritingTools.sharedInstance scheduleShowAffordanceForSelectionRect:selectionRect ofView:m_view.getAutoreleased() forDelegate:(NSObject<WTWritingToolsDelegate> *)m_view.getAutoreleased()];
```
```suggestion
        [WTWritingTools.sharedInstance scheduleShowAffordanceForSelectionRect:selectionRect ofView:m_view.getAutoreleased() forDelegate:(id<WTWritingToolsDelegate>)m_view.getAutoreleased()];
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -6490,7 +6483,7 @@ static NSTextAlignment nsTextAlignmentFromTextAlignment(TextAlignment textAlignm
 
     auto view = m_view.get();
-    showSwapCharactersViewRelativeToRectOfView(selectionBoundsInRootView, view.get());
+    [WTWritingTools.sharedInstance showPanelForSelectionRect:selectionBoundsInRootView ofView:view.get() forDelegate:(NSObject<WTWritingToolsDelegate> *)view.get()];
```
```suggestion
    [WTWritingTools.sharedInstance showPanelForSelectionRect:selectionBoundsInRootView ofView:view.get() forDelegate:(id<WTWritingToolsDelegate>)view.get()];
```

rr-codes committed:
[Writing Tools] Refactor Writing Tools type conversion methods to be in their own file
https://bugs.webkit.org/show_bug.cgi?id=275599
rdar://130054970

Reviewed by Aditya Keerthi.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView wantsWritingToolsInlineEditing]):
(-[WKWebView willBeginWritingToolsSession:requestContexts:]):
(-[WKWebView didBeginWritingToolsSession:contexts:]):
(-[WKWebView proofreadingSession:didReceiveSuggestions:processedRange:inContext:finished:]):
(-[WKWebView proofreadingSession:didUpdateState:forSuggestionWithUUID:inContext:]):
(-[WKWebView didEndWritingToolsSession:accepted:]):
(-[WKWebView compositionSession:didReceiveText:replacementRange:inContext:finished:]):
(-[WKWebView writingToolsSession:didReceiveAction:]):
(-[WKWebView _textReplacementSession:updateState:forReplacementWithUUID:]):
(convert): Deleted.
(-[WKWebView _wantsCompleteUnifiedTextReplacementBehavior]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration setWritingToolsBehavior:]):
(-[WKWebViewConfiguration writingToolsBehavior]):
(convertToPlatform): Deleted.
(convertToWeb): Deleted.
(-[WKWebViewConfiguration _setUnifiedTextReplacementBehavior:]): Deleted.
(-[WKWebViewConfiguration _unifiedTextReplacementBehavior]): Deleted.
(convert): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _web_wantsCompleteUnifiedTextReplacementBehavior]):
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h: Added.
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.mm: Added.
(WebKit::convertToPlatformWritingToolsBehavior):
(WebKit::convertToPlatformTextSuggestionState):
(WebKit::convertToPlatformContext):
(WebKit::convertToWebWritingToolsBehavior):
(WebKit::convertToWebTextSuggestionState):
(WebKit::convertToWebAction):
(WebKit::convertToWebSessionType):
(WebKit::convertToWebCompositionSessionType):
(WebKit::convertToWebContext):
(WebKit::convertToWebSession):
(WebKit::convertToWebTextSuggestion):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::selectionDidChange):
(WebKit::WebViewImpl::canHandleSwapCharacters const):
(WebKit::WebViewImpl::handleContextMenuSwapCharacters):
(WebKit::showSwapCharactersViewRelativeToRectOfView): Deleted.
(WebKit::scheduleShowSwapCharactersViewForSelectionRectOfView): Deleted.
(WebKit::webViewCanHandleSwapCharacters): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/280134@main
